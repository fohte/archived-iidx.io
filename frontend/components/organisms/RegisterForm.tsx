import * as React from 'react'
import { Field as FinalField, Form as FinalForm } from 'react-final-form'
import { Button, Form, Input, Label, Message } from 'semantic-ui-react'
import { isEmpty, isLength, matches } from 'validator'

import withSubmitHandling, {
  ComponentExternalProps,
  InjectedProps,
} from '@app/lib/withSubmitHandling'

export interface ExternalProps extends ComponentExternalProps<FormValues> {}

export interface Props extends ExternalProps, InjectedProps {}

export interface FormValues {
  username: string
  displayName: string
}

const validators: {
  [key in keyof FormValues]: (
    value: FormValues[key] | null,
    allValues: FormValues,
  ) => string | undefined
} = {
  username: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your username.'
    }

    if (!matches(value, /^[a-zA-Z_][a-zA-Z0-9_]*$/)) {
      return 'Username may only contain alphanumeric characters or underscores, and cannot begin with a number.'
    }

    if (!isLength(value, { max: 20 })) {
      return 'Username is too long (maximum is 20 characters).'
    }
  },
  displayName: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your display name.'
    }

    if (!isLength(value, { max: 40 })) {
      return 'Display Name is too long (maximum is 40 characters).'
    }
  },
}

const RegisterForm: React.SFC<Props> = ({ handleSubmit }) => (
  <FinalForm
    onSubmit={handleSubmit}
    initialValues={{ username: '', displayName: '' }}
  >
    {({
      handleSubmit: innerHandleSubmit,
      pristine,
      submitting,
      submitError,
      hasSubmitErrors,
      hasValidationErrors,
    }) => (
      <Form onSubmit={innerHandleSubmit} error={hasSubmitErrors}>
        <FinalField name="username" validate={validators.username}>
          {({ input, meta }) => (
            <Form.Field error={!!(meta.touched && meta.error)}>
              <label>Username</label>
              <Input
                type="text"
                icon="at"
                iconPosition="left"
                {...input}
                placeholder="Username"
              />
              {meta.touched &&
                meta.error && (
                  <Label basic color="red" pointing>
                    {meta.error}
                  </Label>
                )}
            </Form.Field>
          )}
        </FinalField>
        <FinalField name="displayName" validate={validators.displayName}>
          {({ input, meta }) => (
            <Form.Field error={!!(meta.touched && meta.error)}>
              <label>Display Name</label>
              <Input
                type="text"
                icon="user"
                iconPosition="left"
                {...input}
                placeholder="Display Name"
              />
              {meta.touched &&
                meta.error && (
                  <Label basic color="red" pointing>
                    {meta.error}
                  </Label>
                )}
            </Form.Field>
          )}
        </FinalField>
        {hasSubmitErrors && <Message error content={submitError} />}
        <Button
          type="submit"
          disabled={submitting || pristine || hasValidationErrors}
          color="teal"
          size="large"
          loading={submitting}
        >
          Sign up
        </Button>
      </Form>
    )}
  </FinalForm>
)

const WrappedNormalRegisterForm = withSubmitHandling<FormValues>()(RegisterForm)

export default WrappedNormalRegisterForm
