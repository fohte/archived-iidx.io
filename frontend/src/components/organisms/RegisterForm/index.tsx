import * as React from 'react'
import { Field as FinalField, Form as FinalForm } from 'react-final-form'
import { isEmpty, isLength, matches } from 'validator'

import Alert from '@app/components/atoms/Alert'
import Button from '@app/components/atoms/Button'
import FormGroup from '@app/components/atoms/FormGroup'
import InputText from '@app/components/atoms/InputText'
import withSubmitHandling, {
  ComponentExternalProps,
  InjectedProps,
} from '@app/lib/withSubmitHandling'

export interface ExternalProps extends ComponentExternalProps<FormValues> {}

export interface Props extends ExternalProps, InjectedProps {}

export interface FormValues {
  username: string
  displayName: string
}

const validators: {
  [key in keyof FormValues]: (
    value: FormValues[key] | null,
    allValues: FormValues,
  ) => string | undefined
} = {
  username: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your username.'
    }

    if (!matches(value, /^[a-zA-Z_][a-zA-Z0-9_]*$/)) {
      return 'Username may only contain alphanumeric characters or underscores, and cannot begin with a number.'
    }

    if (!isLength(value, { max: 20 })) {
      return 'Username is too long (maximum is 20 characters).'
    }
  },
  displayName: (value, _) => {
    if (!value || isEmpty(value)) {
      return 'Please input your display name.'
    }

    if (!isLength(value, { max: 40 })) {
      return 'Display Name is too long (maximum is 40 characters).'
    }
  },
}

const RegisterForm: React.SFC<Props> = ({ handleSubmit }) => (
  <FinalForm
    onSubmit={handleSubmit}
    initialValues={{ username: '', displayName: '' }}
  >
    {({
      handleSubmit: innerHandleSubmit,
      pristine,
      submitting,
      submitError,
      hasSubmitErrors,
      hasValidationErrors,
    }) => (
      <form onSubmit={innerHandleSubmit}>
        <FinalField name="username" validate={validators.username}>
          {({ input, meta }) => (
            <FormGroup
              label="Username"
              error={!!(meta.touched && meta.error)}
              errorMessage={meta.error}
            >
              <InputText
                {...input}
                error={!!(meta.touched && meta.error)}
                placeholder="Username"
              />
            </FormGroup>
          )}
        </FinalField>
        <FinalField name="displayName" validate={validators.displayName}>
          {({ input, meta }) => (
            <FormGroup
              label="Display Name"
              error={!!(meta.touched && meta.error)}
              errorMessage={meta.error}
            >
              <InputText
                {...input}
                error={!!(meta.touched && meta.error)}
                placeholder="Display Name"
              />
            </FormGroup>
          )}
        </FinalField>
        {hasSubmitErrors && <Alert>{submitError}</Alert>}
        <Button
          type="submit"
          disabled={submitting || pristine || hasValidationErrors}
          loading={submitting}
        >
          Register
        </Button>
      </form>
    )}
  </FinalForm>
)

const WrappedNormalRegisterForm = withSubmitHandling<FormValues>()(RegisterForm)

export default WrappedNormalRegisterForm
