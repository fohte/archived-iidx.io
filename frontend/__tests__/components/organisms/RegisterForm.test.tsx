import * as React from 'react'
import * as renderer from 'react-test-renderer'

import RegisterForm from '@app/components/organisms/RegisterForm'

describe('RegisterForm', () => {
  it('renders correctly', () => {
    const component = renderer.create(<RegisterForm />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
