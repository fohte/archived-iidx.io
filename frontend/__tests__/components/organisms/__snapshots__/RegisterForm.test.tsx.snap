// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm renders correctly 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label>
      Username
    </label>
    <div
      className="ui left icon input"
    >
      <input
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <i
        aria-hidden="true"
        className="at icon"
      />
    </div>
  </div>
  <div
    className="field"
  >
    <label>
      Display Name
    </label>
    <div
      className="ui left icon input"
    >
      <input
        name="displayName"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Display Name"
        type="text"
        value=""
      />
      <i
        aria-hidden="true"
        className="user icon"
      />
    </div>
  </div>
  <button
    className="ui teal large disabled button"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
    type="submit"
  >
    Sign up
  </button>
</form>
`;
